Issue #2483 - Add backendtlspolicy API Documentation (#2547)
kubernetes-sigs/gateway-api: site-src/api-types/backendtlspolicy.md

of the connection from the Gateway to a backend pod/s via the Service API object.

## Background

`BackendTLSPolicy` specifically addresses the configuration of TLS in order to convey HTTPS from the Gateway
dataplane to the backend.  This is referred to as "backend TLS termination" and enables the Gateway to know
how to connect to a backend pod that has its own certificate.  

While there are other API objects provided for TLS to be configured for **passthrough** and **edge** termination,
this API object allows users to specifically configure **backend** TLS termination.  For more information on TLS
configuration in Gateway API, see [TLS Configuration](/guides/tls/).

![Image showing three Termination Types](/images/tls-termination-types.png)

BackendTLSPolicy is a Direct [PolicyAttachment](/reference/policy-attachment/) without defaults or overrides,
applied to a Service that accesses a backend, where the BackendTLSPolicy resides in the same namespace as the
Service to which it is applied. The BackendTLSPolicy and the Service must reside in the same namespace in order
to prevent the complications involved with sharing trust across namespace boundaries.  

> One can reuse the same BackendTLSPolicy for all the different Routes that might point to a referenced Service.

Reviewer: Does this imply "all Routes [attached to a Gateway with support for BackendTLSPolicy] that point to a referenced Service should/must respect a configured policy"?

I think this could be slightly more clear on whether respecting this configuration is optional ("one can reuse") or expected in scenarios with multiple Routes.

Reviewer 2: Good suggestion @mikemorris, maybe something like this? Once this graduates to standard channel, I think "Should" will become "Must" here.
```suggestion
All Gateway API Routes that point to a referenced Service should respect a configured BackendTLSPolicy.
```

Author: Agreed. Thanks!

## Spec

The specification of a [BackendTLSPolicy][backendtlspolicy] consists of:

- [TargetRef][targetRef] - Defines the targeted API object of the policy.  Only Service is allowed.
- [TLS][tls] - Defines the configuration for TLS, including hostname, CACertRefs, and WellKnownCACerts.
- [Hostname][hostname] - Defines the Server Name Indication (SNI) that the Gateway uses to connect to the backend.
- [CACertRefs][caCertRefs] - Defines one or more references to objects that contain PEM-encoded TLS certificates,
which are used to establish a TLS handshake between the Gateway and backend Pod.  Either CACertRefs or WellKnownCACerts
may be specified, but not both.
- [WellKnownCACerts][wellKnownCACerts] - Specifies whether system CA certificates may be used in the TLS
handshake between the Gateway and backend Pod.  Either CACertRefs or WellKnownCACerts may be specified, but not both.

> ![Image showing BackendTLSPolicy objects](images/backendtlspolicy-api.png)

Reviewer 2: This diagram is a bit confusing to me. Maybe it's just that it's hard to distinguish the field name with the field type. I'd consider omitting it for now and adding it in a follow up. For any diagrams that aren't just mermaid.js, is it possible to share a link to a source file that others can copy to modify?

Author: I'm going to share some modified mermaid for this diagram.  I can still omit it if you think it doesn't add any value.  There is one other diagram here (TLS Termination Types) that exists already in the repo.  I can try to find the original source.

Author: @robscott I also just updated the mkdocs.yaml and alphabetized the API types there.  I didn't want BackendTLSPolicy to be the first in the list, so I added a suboption called Policy, where we can put more Policies in the future.

The following illustrates a BackendTLSPolicy that configures TLS for a Service serving a backend:
![HTTPRoute with backendTLSPolicy](/images/httproute-with-backend-tls-policy.png)

```mermaid
flowchart LR
    client(["client"])
    gateway["Gateway"]
    style gateway fill:#02f,color:#fff
    httproute["HTTP<BR>Route"]
    style httproute fill:#02f,color:#fff
    service["Service"]
    style service fill:#02f,color:#fff
    pod1["Pod"]
    style pod1 fill:#02f,color:#fff
    pod2["Pod"]
    style pod2 fill:#02f,color:#fff
    client -.->|HTTP <br> request| gateway
    gateway --> httproute
    httproute -.->|BackendTLSPolicy|service
    service --> pod1 & pod2
```

### Targeting backends

> A BackendTLSPolicy targets a backend Pod (or set of Pods) via a`TargetRef` to a Service.  This TargetRef is a

Reviewer 2: Would recommend without backticks, but that's non-blocking. In either case, looks like another space was needed here.
```suggestion
A BackendTLSPolicy targets a backend Pod (or set of Pods) via a TargetRef to a Service.  This TargetRef is a
```

required object reference that specifies a Service by its Name, Kind (Service), and optionally its Namespace and Group.
TargetRef identifies the Service for which your HTTPRoute requires TLS.

!!! info "Restrictions"

> - Cross-namespace certificate references are not allowed.

Reviewer 2: Hard to tell without working deploy preview, but I think you may need a new line + indentation here to match what you have below. 
```suggestion

    - Cross-namespace certificate references are not allowed.
```

TargetRef identifies the Service for which your HTTPRoute requires TLS.

!!! info "Restrictions"

    - Cross-namespace certificate references are not allowed.

### BackendTLSPolicyConfig

A BackendTLSPolicyConfig is the specification for the BackendTLSPolicy and defines the configuration for TLS,
including hostname (for server name indication) and certificates.

#### Hostname

Hostname defines the server name indication (SNI) the Gateway should use in order to connect to the backend, and must
match the certificate served by the backend pod. A hostname is the fully qualified domain name of a network host, as
defined by [RFC 3986][rfc-3986]. Note the following deviations from the “host” part of the URI as defined in the RFC:

- IP addresses are not allowed.

Also note:

!!! info "Restrictions"

    - Wildcard hostnames are not allowed.

#### Certificates

The BackendTLSPolicyConfig must contain a certificate reference of some kind, and contains two ways to configure the
certificate to use for backend TLS, CACertRefs and WellKnownCACerts.  Only one of these may be used per
BackendTLSPolicyConfig.

> ##### CaCertRefs

Reviewer 2: Potential follow up - this is currently styled in all caps (https://deploy-preview-2547--kubernetes-sigs-gateway-api.netlify.app/api-types/backendtlspolicy/). It might be better to just go with bold text here instead of a heading to avoid that styling.
